### PR TITLE
Add Compose Onboard Feature

### PR DESCRIPTION
**Pull request by oberien**
This PR adds the Compose Onboard feature, which allows composing directly on the keyboard instead of requiring userland composing (like `.XCompose` on Linux).

In short, after pressing the `KC_COMPOSE_ONBOARD` key, a sequence of keys can be pressed, which can be mapped to actions in code. This is somewhat similar to the Leader Key feature with the main distinction that the leader key is time-based, while this is only input-based.

I used the Leader Key feature implementation as a guideline and I think I have added code / documentation everywhere where the Leader Key appears (according to grep ;) ).

I use it in [my own keymap](https://github.com/oberien/ergodox-ez-neo) (the compose part can be seen [here](https://github.com/oberien/ergodox-ez-neo/blob/master/src/compose.c)).

## Checklist

- [x] My code follows the code style of this project.
    * hopefully, if not, please tell me so I can change it
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
    * I don't know how this can be tested via unit tests. Also I don't think I saw any test file for the Leader Key feature.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

Edit: One possible problem is that the mapping dictionary requires quite a bit of space, increasing the binary size. For example just the `shrug` mapping in the documentation example requires close to 200 bytes.

### CODE REVIEW

**Comment by drashna** [comment]
Just a heads up, the UCIS method does something like this, but explicitly for unicode type stuff.  It's not as customizable, where this is somewhere between UCIS and leader keys.

**Comment by oberien** [comment]
I rebased it onto master to resolve merge conflicts in `common_features.mk` (order changed) and `docs/ja/_summary.md` (order changed in #8508). For the Japanese summary, I don't know if I put this feature in the correct category as I don't understand Japanese. However, I think that it is similar enough to the leader key, that it should also fit into "Advanced key code" as Google Translator translated it.

Is there anything else missing that blocks a review and possible merge?

File: docs/keycodes.md [hunk 1]
```diff
@@ -347,6 +347,14 @@ See also: [Leader Key](feature_leader_key.md)
 |---------|------------------------|
 |`KC_LEAD`|Begins a leader sequence|
 
+## Compose Onboard :id=compose-onboard
+
+See also: [Compose Onboard](feature_compose_onboard.md)
+
+|Key                 |Aliases    |Description                              |
+|--------------------|-----------|-----------------------------------------|
+|`KC_COMPOSE_ONBOARD`|`KC_COMPOB`|Begins Composing directly on the keyboard|
```
Comment (noroadsleft) [inline docs/keycodes.md:356]:
```suggestion
|`KC_COMPOSE_ONBOARD`|`KC_CMOB`|Begins Composing directly on the keyboard|
```

We like these keycode aliases to be seven characters or less, when possible.

Also, this entire section should be moved up in the document to be between the Bluetooth and Dynamic Macros sections, as this page is organized alphabetically (with the exception of the Basic and Quantum sections).

File: docs/reference_glossary.md [hunk 2]
```diff
@@ -78,6 +78,11 @@ A feature that allows you to tap the leader key followed by a sequence of 1, 2,
 
 * [Leader Key Documentation](feature_leader_key.md)
 
+## Compose Onboard
+Allows composing directly on the keyboard. After pressing the compose onboard key, the following keystrokes trigger a preconfigured action like pressing other keys, inserting a character, or any other quantum features.
+
+* [Compose Onboard Documentation](feature_compose_onboard.md)
+
```
Comment (noroadsleft) [inline docs/reference_glossary.md:85]: I don't know that this needs to be added to the glossary, but if its decided that it should be so, this section should move up between Compile and Dvorak, as this page is organized alphabetically.

File: quantum/quantum_keycodes.h [hunk 3]
```diff
@@ -123,6 +123,11 @@ enum quantum_keycodes {
     KC_LEAD,
 #endif
 
+#ifdef COMPOSE_ONBOARD_ENABLE
+    KC_COMPOSE_ONBOARD,
+    KC_COMPOB = KC_COMPOSE_ONBOARD,
+#endif
+
```
Comment (noroadsleft) [inline quantum/quantum_keycodes.h:130]: 1. The `KC_COMPOSE_ONBOARD` keycode should be added near the end of the `quantum_keycodes` enum, just before `SAFE_RANGE` on Line 520, so as not to break compatibility with [VIA Configurator](https://caniusevia.com/).
2 The `KC_COMPOB` alias should be created using a `#define` statement around Line 814, as `#define KC_CMOB KC_COMPOSE_ONBOARD` if you've applied the suggestion I gave above.

File: docs/keycodes.md (hunk 1 above)
Comment (fauxpark) [reply docs/keycodes.md:356]: IMO Quantum keycodes should not have the `KC_` prefix either, as that is reserved for the basic keycode set.

**Comment by drashna** [comment]
Honestly, I am opposed to this change.  

It feels like it is very much what UCIS does already. 
And even if that wasn't the case, if #6580 gets merged, then I believe that this will be duplicating functionality. 

In either case, it feels redundant.  I think the better direction would be to get #6580 merged in.

That said, we absolutely appreciate the time and effort that went into this.  Even if it doesn't get merged, we're happy to see the contribution!

**Comment by tzarc** [comment]
Retargeted to `develop` due to core changes. Discussion regarding the relevance is still outstanding.

**Comment by drashna** [comment]
There is a merge conflict here, now.  Needs a rebase.

**Comment by oberien** [comment]
With #6580 now being merged, I don't think there is a reason left for this PR to stay open. Thanks for your effort in reviewing and rebasing it.

**Pull request closed without merging**